### PR TITLE
Issue #3300: fixed ERROR whitespace after annotation

### PR DESCRIPTION
**Pull request by sergileon**
Issue #3300 fixed

### CODE REVIEW

File: src/xdocs/config_whitespace.xml [hunk 1]
```diff
@@ -961,7 +961,13 @@ import static java.math.BigInteger.ZERO;
...
           href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#INDEX_OP">INDEX_OP</a>
           tokens specially from other tokens. Actually it is checked that there is
-          no whitespace before this tokens, not after them.
+          no whitespace before this tokens, not after them. Annotation before
+          <a
+          href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#ARRAY_DECLARATOR">ARRAY_DECLARATOR</a>
+          and
+          <a
+          href="apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#INDEX_OP">INDEX_OP</a>
+          will be ignored.
```
Comment (romani) [inline src/xdocs/config_whitespace.xml]: Please provide example for this case, users like examples. Nobody have deep understanding of our AST structures.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java [hunk 2]
```diff
@@ -136,12 +138,15 @@ public void setAllowLineBreaks(boolean allowLineBreaks) {
     public void visitToken(DetailAST ast) {
         final DetailAST whitespaceFollowedAst = getWhitespaceFollowedNode(ast);
 
-        final int whitespaceColumnNo = getPositionAfter(whitespaceFollowedAst);
-        final int whitespaceLineNo = whitespaceFollowedAst.getLineNo();
+        if (whitespaceFollowedAst.getNextSibling() == null
+                || whitespaceFollowedAst.getNextSibling().getType() != TokenTypes.ANNOTATIONS) {
+            final int whitespaceColumnNo = getPositionAfter(whitespaceFollowedAst);
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: Whitespace package is not yet enforced pitest coverge in CI , please generate report and place screenshot that you covered all your changes

Comment (sergileon) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: [Pitest report](https://sergileon.github.io/3300/PITEST/index.html)

**Comment by sergileon** [comment]
[Regression testing report](https://sergileon.github.io/3300/index.html)

File: src/xdocs/config_whitespace.xml (hunk 1 above)
Comment (sergileon) [reply src/xdocs/config_whitespace.xml]: I added examples and link for users

File: src/xdocs/config_whitespace.xml [hunk 3]
```diff
@@ -1060,6 +1068,15 @@ import static java.math.BigInteger.ZERO;
     &lt;property name=&quot;allowLineBreaks&quot; value=&quot;false&quot;/&gt;
 &lt;/module&gt;
         </source>
+        <p>
+         If the annotation is between the type and the array, the check will skip:
+        </p>
+        <source>
+public void a(final char @AnnotationAfterTest [] a) {}//Correct
```
Comment (romani) [inline src/xdocs/config_whitespace.xml]: please put there `@NotNull` to be close to real life.
All cases from below please remove, they are too syntactic.
"//Correct" ==> "//No violation"

File: src/xdocs/config_whitespace.xml [hunk 4]
```diff
@@ -1060,6 +1068,15 @@ import static java.math.BigInteger.ZERO;
     &lt;property name=&quot;allowLineBreaks&quot; value=&quot;false&quot;/&gt;
 &lt;/module&gt;
         </source>
+        <p>
+         If the annotation is between the type and the array, the check will skip:
```
Comment (romani) [inline src/xdocs/config_whitespace.xml]: "the check will skip" ==> "the check will skip validation for space(s)"

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java [hunk 5]
```diff
@@ -51,6 +51,9 @@
  * {@link TokenTypes#INDEX_OP INDEX_OP}
  * specially from other tokens. Actually it is checked that there is
  * no whitespace before this tokens, not after them.
+ * {@link TokenTypes#ANNOTATIONS ANNOTATIONS}
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: Check does not ignore annotations, it is ignoring spaces.

**Comment by romani** [comment]
@sergileon , please rebase on our master HEAD to fix eclipse problem .

**Comment by sergileon** [comment]
@romani , I used windows to generate a report and I catch an error: ` Filename too long`.
My command for run:
`groovy diff.groovy --localGitRepo D:\project\check\checkstyle --baseBranch master --patchBranch whitespaceAfterAnnotanion2 --config my_check.xml --listOfProjects projects-for-circle.properties --shortFilePaths true`
How use command `shortFilePaths`?

File: src/xdocs/config_whitespace.xml (hunk 4 above)
Comment (sergileon) [reply src/xdocs/config_whitespace.xml]: fixed

File: src/xdocs/config_whitespace.xml (hunk 3 above)
Comment (sergileon) [reply src/xdocs/config_whitespace.xml]: fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java (hunk 5 above)
Comment (sergileon) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: fixed

**Comment by romani** [comment]
> How use command shortFilePaths?

please looks at git history, there should be links to issues, all details in issues.
@rnveach , please assist, as you the only with Windows.

**Comment by sergileon** [comment]
Update report [https://sergileon.github.io/3300/index.html](https://sergileon.github.io/3300/index.html)

File: src/xdocs/config_whitespace.xml (hunk 3 above)
Comment (Vladlis) [reply src/xdocs/config_whitespace.xml]: @sergileon 
not fixed, please read attentively - you need to keep only one example:
```java
public void a(final char @NotNull [] a) {} // No violation
```

**Comment by sergileon** [comment]
@romani , Why ci failed? I looked at the logs and didn't see the error.

**Comment by romani** [comment]
@sergileon , please rebase, I did fix for this type of problem yesterday.

File: src/xdocs/config_whitespace.xml (hunk 3 above)
Comment (sergileon) [reply src/xdocs/config_whitespace.xml]: @Vladlis  fixed

**Comment by rnveach** [comment]
@sergileon please make comment when you have finished what we asked. We don't get notifications on new commits, so I assume this PR went unnoticed.

**Comment by rnveach** [comment]
@romani Please finish your review since you started.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java [hunk 6]
```diff
@@ -67,6 +70,12 @@
  *     &lt;property name="allowLineBreaks" value="false"/&gt;
  * &lt;/module&gt;
  * </pre>
+ * <p>
+ * If the annotation is between the type and the array, the check will skip validation for spaces:
+ * </p>
+ * <pre>
+ * public void a(final char @NotNull [] param) {} // No violation
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: "a" ==> "foo".
Do not show bad practices to user, it is even hard to read.

File: src/xdocs/config_whitespace.xml [hunk 7]
```diff
@@ -1060,6 +1069,12 @@ import static java.math.BigInteger.ZERO;
     &lt;property name=&quot;allowLineBreaks&quot; value=&quot;false&quot;/&gt;
 &lt;/module&gt;
         </source>
+        <p>
+         If the annotation is between the type and the array, the check will skip validation for spaces:
+        </p>
+        <source>
+public void a(final char @NotNull [] param) {} // No violation
```
Comment (romani) [inline src/xdocs/config_whitespace.xml]: "a" ==> "foo"

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java (hunk 6 above)
Comment (sergileon) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java]: @romani , fixed.

File: src/xdocs/config_whitespace.xml (hunk 7 above)
Comment (sergileon) [reply src/xdocs/config_whitespace.xml]: @romani , fixed.

**Pull request merged**